tracing/ops_tracing/_api.py

```diff
@@ -0,0 +1,223 @@
+# Copyright 2025 Canonical Ltd.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""The tracing API for the charms."""
+
+from __future__ import annotations
+
+import dataclasses
+import logging
+
+import ops
+
+from ._const import Config
+from .vendor.charms.certificate_transfer_interface.v1.certificate_transfer import (
+    CertificateTransferRequires,
+)
+from .vendor.charms.tempo_coordinator_k8s.v0.tracing import (
+    AmbiguousRelationUsageError,
+    ProtocolNotRequestedError,
+    RelationInterfaceMismatchError,
+    RelationNotFoundError,
+    RelationRoleMismatchError,
+    TracingEndpointRequirer,
+)
+
+# Databag operations can be trivially re-coded in pure Python:
+# - we'd get rid of pydantic dependency
+# - we'd avoid init-time relation listing and databag update (tracing)
+# - it's trivial to read the relation data bags, see below.
+#
+#
+# certificate_transfer:
+# - client doesn't write to the databag
+# - client reads this from the databag:
+# ```
+# {"certificates": list[str]}
+# ```
+#
+#
+# tracing:
+# - client leader sets own app data on every relation:
+# ```
+# {"receivers": ["otlp_http"]}
+# ```
+#
+# - client reads this from remote_app_data:
+# ```
+# {"receivers": [
+#     {
+#         "protocol": {"name": "otlp_http", "type": "http"},
+#         "url": "http//somewhere:4318/v1/traces",
+#     },
+#     ...,
+# ]}
+# ```
+
+
+class Tracing(ops.Object):
+    """Tracing service."""
+
+    def __init__(
+        self,
+        charm: ops.CharmBase,
+        tracing_relation_name: str,
+        ca_relation_name: str | None = None,
+        ca_data: str | None = None,
+    ):
+        """Initialise the tracing service.
```

Comment:
Why does this doc string not show up in the docs?
(The class is listed, but not any of its methods).
Is it because the Tracing symbol is re-exported?